single download for alpine sources

deitch opened the pull request:
This update does a few things.

* rather than `curl` getting for the commit for each package, it clones the repo just once, to a tempdir, and then checks out different commits for each package (main change).
* provides an option to use `-g <gitdir>` in case you already have it checked out, in which case it just copies to the tmpdir instead of cloning it; this is faster and useful for testing or including in other workflows.
* `set -e` for the script, so any errors fail immediately
* simplify some of the `if`/`then`/`else` statements

eriknordmark reviewed (changes requested):
I tried running this with the same arguments as I've run the current script (to compare the resulting sources) but script silently exit 1!
When I run with -v I get:
tools/get-alpine-pkg-source.sh -s alpine-sources4 -u urls -v 9.9.0-kvm-amd64
outdir: alpine-sources4
retrieving build_config for 9.9.0-kvm-amd64
processing docker.io/linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7
nordmark@bobo:~/9.9.0/eve$ echo $?
1

I thought there was no change to the usage - just the addition of -g.

deitch commented:
I also added `set -e` so it will exist in the case of failure. Did some command fail? Try running it with `sh -x` and see what it does?

eriknordmark commented:
nordmark@bobo:~/fixes/eve$ bash -x tools/get-alpine-pkg-source.sh -s alpine-sources4 -u urls -v 9.9.0-kvm-amd64
+ set -e
+ verbose=
+ tags=
+ evedir=
+ urlfile=
+ gitdir=
+ outdir=/tmp/64814
+ quiet=
+ getopts e:vt:u:s:g:q o
+ case "$o" in
+ outdir=alpine-sources4
+ getopts e:vt:u:s:g:q o
+ case "$o" in
+ urlfile=urls
+ getopts e:vt:u:s:g:q o
+ case "$o" in
+ verbose=1
+ getopts e:vt:u:s:g:q o
+ shift 5
+ '[' 1 == 0 ']'
+ '[' 1 -gt 1 ']'
+ '[' -d alpine-sources4 ']'
+ '[' -n '' ']'
++ pwd
+ startdir=/home/nordmark/fixes/eve
+ mkdir -p alpine-sources4
+ cd alpine-sources4
+ '[' -n 1 ']'
+ echo 'outdir: alpine-sources4'
outdir: alpine-sources4
+ '[' 1 == 1 ']'
+ VERSION=9.9.0-kvm-amd64
+ '[' -n 1 ']'
+ echo 'retrieving build_config for 9.9.0-kvm-amd64'
retrieving build_config for 9.9.0-kvm-amd64
+ docker run lfedge/eve:9.9.0-kvm-amd64 build_config
++ grep image: build_config.out
++ awk '{print $2}'
+ tags1='docker.io/lfedge/eve-kernel:9.9.0-amd64
docker.io/lfedge/eve-rngd:9.9.0-amd64
docker.io/linuxkit/sysctl:v0.5
docker.io/linuxkit/modprobe:v0.5
docker.io/lfedge/eve-storage-init:9.9.0-amd64
docker.io/lfedge/eve-kdump:9.9.0-amd64
docker.io/lfedge/eve-pillar:9.9.0-amd64
docker.io/lfedge/eve-measure-config:9.9.0-amd64
docker.io/lfedge/eve-newlog:9.9.0-amd64
docker.io/lfedge/eve-edgeview:9.9.0-amd64
docker.io/lfedge/eve-debug:9.9.0-amd64
docker.io/lfedge/eve-wwan:9.9.0-amd64
docker.io/lfedge/eve-wlan:9.9.0-amd64
docker.io/lfedge/eve-guacd:9.9.0-amd64
docker.io/lfedge/eve-pillar:9.9.0-amd64
docker.io/lfedge/eve-vtpm:9.9.0-amd64
docker.io/lfedge/eve-watchdog:9.9.0-amd64
docker.io/lfedge/eve-xen-tools:9.9.0-amd64'
++ awk '/^init:/ {pr=1} /^onboot:/ {pr=0} /^-/ {if (pr) {print $2}}'
+ tags2='docker.io/linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7
docker.io/linuxkit/runc:f01b88c7033180d50ae43562d72707c6881904e4
docker.io/linuxkit/containerd:de1b18eed76a266baa3092e5c154c84f595e56da
docker.io/linuxkit/getty:c9d5afa9a61ac907904090643e946874ff6bf07c
docker.io/linuxkit/memlogd:v0.5
docker.io/lfedge/eve-dom0-ztools:9.9.0-amd64
docker.io/lfedge/eve-grub:9.9.0-amd64
docker.io/lfedge/eve-fw:9.9.0-amd64
docker.io/lfedge/eve-xen:9.9.0-amd64
docker.io/lfedge/eve-gpt-tools:9.9.0-amd64
docker.io/lfedge/eve-kexec:9.9.0-amd64'
+ tags='docker.io/linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7
docker.io/linuxkit/runc:f01b88c7033180d50ae43562d72707c6881904e4
docker.io/linuxkit/containerd:de1b18eed76a266baa3092e5c154c84f595e56da
docker.io/linuxkit/getty:c9d5afa9a61ac907904090643e946874ff6bf07c
docker.io/linuxkit/memlogd:v0.5
docker.io/lfedge/eve-dom0-ztools:9.9.0-amd64
docker.io/lfedge/eve-grub:9.9.0-amd64
docker.io/lfedge/eve-fw:9.9.0-amd64
docker.io/lfedge/eve-xen:9.9.0-amd64
docker.io/lfedge/eve-gpt-tools:9.9.0-amd64
docker.io/lfedge/eve-kexec:9.9.0-amd64 docker.io/lfedge/eve-kernel:9.9.0-amd64
docker.io/lfedge/eve-rngd:9.9.0-amd64
docker.io/linuxkit/sysctl:v0.5
docker.io/linuxkit/modprobe:v0.5
docker.io/lfedge/eve-storage-init:9.9.0-amd64
docker.io/lfedge/eve-kdump:9.9.0-amd64
docker.io/lfedge/eve-pillar:9.9.0-amd64
docker.io/lfedge/eve-measure-config:9.9.0-amd64
docker.io/lfedge/eve-newlog:9.9.0-amd64
docker.io/lfedge/eve-edgeview:9.9.0-amd64
docker.io/lfedge/eve-debug:9.9.0-amd64
docker.io/lfedge/eve-wwan:9.9.0-amd64
docker.io/lfedge/eve-wlan:9.9.0-amd64
docker.io/lfedge/eve-guacd:9.9.0-amd64
docker.io/lfedge/eve-pillar:9.9.0-amd64
docker.io/lfedge/eve-vtpm:9.9.0-amd64
docker.io/lfedge/eve-watchdog:9.9.0-amd64
docker.io/lfedge/eve-xen-tools:9.9.0-amd64'
+ OCPAIRS=/tmp/ocpairs.64814
+ '[' -n '' ']'
+ for TAG in ${tags}
+ '[' -n 1 ']'
+ echo 'processing docker.io/linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7'
processing docker.io/linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7
++ mktemp /tmp/installed.64814.XXXXXX
+ installed=/tmp/installed.64814.DJBQXb
++ mktemp /tmp/ocpairs.64814.XXXXXX
+ ocpairs=/tmp/ocpairs.64814.O17Pjo
++ docker create docker.io/linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7
+ c=
nordmark@bobo:~/fixes/eve$

deitch commented:
I think you meant backticks instead of single-quotes 😄

deitch commented:
It looks like `docker create docker.io/linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7` returned non-zero, so it failed. We can add some better handling around that.

deitch commented:
Yup, you need to give it a command.

```
$ docker create docker.io/linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7
Unable to find image 'linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7' locally
8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7: Pulling from linuxkit/init
90b2401061ec: Pull complete
d08d7bfead38: Pull complete
eb09e7dd8c4c: Pull complete
a16bac6dbdf9: Pull complete
748940886e61: Pull complete
0aae814b63c9: Pull complete
Digest: sha256:40b01549bf87ddf7afc157ad1c0197ae7cc4dffb3014e72b43a80b25c6173f78
Status: Downloaded newer image for linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7
Error response from daemon: No command specified

$ echo $?
1
```

deitch commented:
I updated it to not send stderr to `/dev/null`, so if it has an issue, it will print it out.

I also had it always send a dummy command. 

This should help.

eriknordmark commented:
> This should help.

Got a bit further, then died on 
processing docker.io/linuxkit/init:8f1e6a0747acbbb4d7e24dc98f97faa8d1c6cec7
Error: No such container:path: 3cd999eaea447b4d1fa3b03afe36a437867f1db178e266271465a61fd5f42297:/lib/apk/db/installed

Can you test it using e.g., 
tools/get-alpine-pkg-source.sh -s alpine-sources4 -u urls -v 9.9.0-kvm-amd64
so I don't have to repeat running that with your patch?

eriknordmark commented:
There is another bug even after I remove the set -e; it doesn't copy anything when outdir is a relative path with errors of the form
cp: cannot create directory 'alpine-sources4/zstd.1.5.2-r1.22b9167f4898653b61c7a8f289f631ac17f83740/': No such file or directory

Fix by changing L173 to
    dstdir="${origin}.${version}.${commit}"

eriknordmark commented:
Also, there are files missing compared to the old version of the script (plus some additional files).
An example of missing files are in alpine-baselayout.348653a9ba0701e8e968b3344e72313a9ef334e4 where the protocols and services files in APKBUILD use macros and refer to a http download. Those do not get included with the new version of the script.

The additional files appear to be files which are in github but are not listed in sources in APKBUILD. I guess there is no harm in including them.

But in general, need to run the old and new versions of the script and verify that things are not missing in the new version. The new version of the script outputs 
Missing/bad 29 files
which indicates it has a problem with at least the http entries in sources.

deitch commented:
> cp: cannot create directory 'alpine-sources4/zstd.1.5.2-r1.22b9167f4898653b61c7a8f289f631ac17f83740/': No such file or directory

Fixed

> Missing/bad 29 files

Fixed as well, except if the file actually is missing. I show 6 files missing; we will figure out which.

deitch commented:
Updated it to report what the issue with missing or badflies is, which file and package, and report those at the end.

deitch commented:
Latest run has the following list:

```
bad-content:ncurses.6.2_p20210612-r0.53504bd606f7d8bca692c5a0e46b1e01c13cecdc:ncurses-6.2-20210612.tgz
bad-content:ncurses.6.3_p20220521-r0.bce94233ff8618dcf52d30076710d791798b424c:ncurses-6.3-20220521.tgz 
mismatched-sh512:pixman.0.40.0-r3.8.e1f6d0b0f4cca3d659519153071083ca92e3594:04f662a3ee0dbab1b4755a6faf491751f6f29627.patch 
missing:wireless-tools.30_pre9-r1.5.b043bf056b9af4ae6a50428a460350fa4942bc3:wireless_tools.30.pre9.tar.gz
bad-content:zlib.1.2.11-r3.388.a4fb3640f8ccbd18e105df3ad741dca4247e1:zlib-1.2.11.tar.gz
bad-content:zlib.1.2.12-r3.57.ce38bde7ce42964b664c137935cf2de803ac44:zlib-1.2.12.tar.gz
```

deitch commented:
Did a revamp. Now, instead of reading `build_config` (which the build time did anyways), it extracts the rootfs and pulls out the `'*lib/apk/db/installed` files and their contents and processes them.

Works rather nicely, actually. Getting the same output for both `-e dir` and the tag you suggested. Same missing files:

```
404-not-found:ncurses.6.2_p20210612-r0.53504bd606f7d8bca692c5a0e46b1e01c13cecdc:ncurses-6.2-20210612.tgz
404-not-found:ncurses.6.3_p20220521-r0.bce94233ff8618dcf52d30076710d791798b424c:ncurses-6.3-20220521.tgz
mismatched-sh512:pixman.0.40.0-r3.8e1f6d0b0f4cca3d659519153071083ca92e3594:04f662a3ee0dbab1b4755a6faf491751f6f29627.patch
missing:wireless-tools.30_pre9-r1.5b043bf056b9af4ae6a50428a460350fa4942bc3:wireless_tools.30.pre9.tar.gz
404-not-found:zlib.1.2.11-r3.388a4fb3640f8ccbd18e105df3ad741dca4247e1:zlib-1.2.11.tar.gz 
404-not-found:zlib.1.2.12r3.57ce38bde7ce42964b664c137935cf2de803ac44:zlib-1.2.12.tar.gz
``` 

Take a look @eriknordmark ?

deitch commented:
@eriknordmark I ran a diff. Generated using the existing one and post-this PR. Then I did a `find -exec ls -l {} \:` for each, saving it to a file, and removing the date (which is irrelevant). Then I did a diff between them.

There are some files in the new that are not old, as you rightly pointed out. But the new does not miss any file from the old, and anything that is in the old has the same file size in the new.

eriknordmark commented:
Looking through my notes for the problematic references from APKBUILD from a few weeks ago I see a similar list.
Problematic files:
1. https://downloads.sourceforge.net/project/expat/expat/2.4.8/expat-2.4.8.tar.bz2
    Found exact file (same sha512) at
https://sourceforge.net/projects/expat/files/expat/2.4.8/expat-2.4.8-RENAMED-VULNERABLE-PLEASE-USE-2.5.0-INSTEAD.tar.bz2

2. https://invisible-mirror.net/archives/ncurses/current/ncurses-6.2-20210612.tgz
    I can not find a version with the same sha256. Close version is:
https://invisible-mirror.net/archives/ncurses/ncurses-6.2.tar.gz

3. http://www.hpl.hp.com/personal/Jean_Tourrilhes/Linux/wireless_tools.30.pre9.tar.gz
    Exact file (same sha256) exists at
https://hewlettpackard.github.io/wireless-tools/wireless_tools.30.pre9.tar.gz

4. https://zlib.net/zlib-1.2.11.tar.gz
    Exact file (same sha256) exists at
https://github.com/fabianthiery/zlib-1.2.11/blob/main/zlib-1.2.11.tar.gz

5. https://zlib.net/zlib-1.2.12.tar.gz
    Approximate file at
https://github.com/madler/zlib/archive/refs/tags/v1.2.12.tar.gz

6. https://gitlab.freedesktop.org/pixman/pixman/-/commit/04f662a3ee0dbab1b4755a6faf491751f6f29627.patch
    This file was included but it has a different sha512 that what is in
the APKBUILD file, which is odd. The file is a patch which only affects
ARM builds.

Your list is slightly different than the above, but my run was a month ago so the versions we use in EVE might have changed

eriknordmark reviewed (approved):
Script is ok; verified it is the same output as the old script.

deitch commented:
> but my run was a month ago so the versions we use in EVE might have changed

I believe that is the case. I ran the script with `set -x` and walked through the many thousands of lines of output. Those are the 6.

If that is the case, I think we can merge this in, as it improves things, and then we can add extra handling for the above.
